### ISSUE TITLE
[GUI] The default theme (and a few others) have problems with Cyrillic text on Windows

### ISSUE THREAD

**Issue by victoryforce** [darktable-org/darktable#13084]
**Describe the bug/issue**

The default theme for the Ukrainian UI (as well as other languages with _Cyrillic_ text) contains text with too wide letters and with large spacing between letters, which looks monospaced.

The problem does not appear only in the "darktable" and "darktable-icons" themes.

**To Reproduce**

1. Launch current Windows darktable build and change the interface language to Ukrainian.
2. Restart darktable. If you do this with an empty configuration directory, then you will get the theme "darktable-elegant-grey" and you can observe the problem. 

**Platform**
* darktable version:  Latest git master build as of the date of the issue.
* OS: Windows

**Additional context**

This is not reproducible on Linux (at least on a regular Ubuntu installation).

**Comment by dterrahe** [comment]
A screenshot helps (from Windows 10)
![image](https://user-images.githubusercontent.com/1549490/205914351-61d8f855-db47-463a-bb49-88574e4e2c4f.png)

It seems that for example module labels (and tooltips, which are harder to screenshot on windows) are not affected, but bauhaus labels are. This may help @Nilvus diagnose the issue, although this could also be a lack of appropriate fonts on windows.

(this probably also helped aggravate #12877)

**Comment by victoryforce** [comment]
> although this could also be a lack of appropriate fonts on windows.

Perhaps, but no fonts have been explicitly removed from Windows recently, and there was no such problem before. I can't remember the exact date to help find the cause, but I used to run dt with an empty configuration for debugging, it immediately started with the Ukrainian language (from system settings) and this problem did not occur.

**Comment by Nilvus** [comment]
If it appears only on elegant themes, so maybe some Roboto fonts missing. I don't know how they handle Cyrillic fonts but they should given that those theme use them since 3 years now and nothing had changed recently.

The only font setting recently done was adding that line, following your good idea @dterrahe:

`font-feature-settings: "tnum";` (line 241)

That line is add spacinf for number only. But maybe that could have unwanted effet on some Cyrillic fonts. Is one of you could test what you have if you remove that line?

**Comment by victoryforce** [comment]
> If it appears only on elegant themes, so maybe some Roboto fonts missing. I don't know how they handle Cyrillic fonts but they should given that those theme use them since 3 years now and nothing had changed recently.

This problem does not occur on darktable 4.0.1. So it can't be a missing font issue because the css file that defines the fonts hasn't changed since then.

> The only font setting recently done was adding that line, following your good idea @dterrahe:
> 
> `font-feature-settings: "tnum";` (line 241)
> 
> That line add spacing for numbers only. But maybe that could have unwanted effet on some Cyrillic fonts. Is one of you could test what you have if you remove that line?

In my test, removing this line does not fix the problem.

**Comment by Nilvus** [comment]
So there's something else. Unfortunately, I can't help more here. If this works great on darktable 4.0.1, that means it would be good if we fix that for 4.2 release (we have 1 week for that, no more). But, as you precise that you can't reproduce on Linux, I can't reproduce so. A git bisect would so be needed to understand when this had been introduced.

**Comment by Mark-64** [comment]
> This problem does not occur on darktable 4.0.1. So it can't be a missing font issue because the css file that defines the fonts hasn't changed since then.

Windows 11, I rebuilt 4.0.1 and the issue is there.
Could it be some updates in Windows fonts?

**Comment by Mark-64** [comment]
In `darktable-elegant-darker.css` (for example), line 71
{
```
  font-family: "Roboto Light", "Roboto",                  /* best case scenario */
               "Segoe UI Light", "Segoe UI",              /* Windows 10 default */
               "SF Pro Display Light", "SF Pro Display",  /* Mac OS X default */
               "Ubuntu Light", "Ubuntu", "IPAPGothic",    /* Ubuntu default */
               "Cantarell",                               /* Gnome default */
                sans-serif;                                /* default default */
}
```
removing `Segoe UI Light` fixes the issue for me.
Definitely a Windows and/or Gtk problem.

@Nilvus what we do?

**Comment by kmilos** [comment]
["Light" is one of the 12 font faces ("tipi") of Segoe UI you see there, and should have Cyrillic support.](https://learn.microsoft.com/en-us/typography/font-list/segoe-ui)

**Comment by victoryforce** [comment]
> Windows 11, I rebuilt 4.0.1 and the issue is there.

I have Windows 10, latest updates, nothing font related installed or uninstalled. When I wrote that the problem does not occur on darktable 4.0.1, I meant the installed __official build__. It is possible that dt 4.0.1 rebuilt now, with the current versions of the libraries in MinGW, would also suffer from this problem (not yet tested though). Then it may indeed look like a regression in some related libraries.

> Actually neither Roboto (any variant) nor Segoe UI Light have cyrillic support on my Windows 11 system.

No Roboto font installed on my system. Regarding "Segoe UI Light" I was not convinced by your screenshot that this font does not contain Cyrillic. Only "Segoe UI" is shown there, but all other typefaces of this font are hidden behind this base font name.

In order to check whether the font contains Cyrillic letters, I use "Character Map" Windows utility. Segoe UI Light actually contains Cyrillic letters.

**Comment by kmilos** [comment]
> It is possible that dt 4.0.1 rebuilt now, with the current versions of the libraries in MinGW

Yep, it wouldn't be the first time a pango/cairo/harfbuzz/gtk update introduced a regression on Windows, but it is not easy to track down...

**Comment by kmilos** [comment]
@victoryforce Maybe try [downloading](https://packages.msys2.org/repos) and replacing with older DLLs of said libraries one by one?

**Comment by Nilvus** [comment]
Another annoying Windows specific issue. I fear that removing Segoe UI Light could break display on other languages for some Windows users. I remember that Segoe fonts were added to allow correct rendering on Windows (on non Cyrillic languages, Cyrillic were not tested). Seems that regarding your comments, removing it would be more a workaround to something broken on some external libraries than a good fix.

**Comment by kmilos** [comment]
> Another annoying Windows specific issue.

We have nightly builds, but not enough testers :(

**Comment by Mark-64** [comment]
Manually downgrading pango in MSYS2 to 
`mingw-w64-ucrt-x86_64-pango-1.48.9-2`
fixes the issue for me, so confirmed it is a regression in pango.
I will try to find the last good version and open an issue upstream.

So for 4.2 we could build the Windows version against an old release of pango, waiting for the fix.

Can someone confirm?

**Comment by Mark-64** [comment]
To me, the last good version of pango is
`mingw-w64-ucrt-x86_64-pango-1.50.9-1`
the one immediately before the current 
`mingw-w64-ucrt-x86_64-pango-1.50.12-1`

Good, the regression is very recent

**Comment by TurboGit** [comment]
@Mark-64 : Nice detective work. Thanks!

**Comment by TurboGit** [comment]
@wpferguson : Please can you record this to be done (using mingw-w64-ucrt-x86_64-pango-1.50.9-1) for the next Windows binary release? TIA.

**Comment by kmilos** [comment]
> which is actually the one immediately before the current

Hm, I see 1.5.10 and 1.5.11 also available here: https://repo.msys2.org/mingw/ucrt64/ so might be nice to narrow this down further, thanks! @Mark-64

**Comment by Mark-64** [comment]
> I see 1.5.10 and 1.5.11 also available here

You're right, I didn't see those, sorry.
.11 is still good, .12 is broken.
So the last good version for building DT is 

`mingw-w64-ucrt-x86_64-pango-1.50.11-1`

**Comment by TurboGit** [comment]
@wpferguson : Actually last good version found by @Mark-64 is **mingw-w64-ucrt-x86_64-pango-1.50.11-1**

**Comment by wpferguson** [comment]
@TurboGit that's the version I have installed and have been using for the weekly builds.  I'll make sure I stay on it until after the release

**Comment by Mark-64** [comment]
Update on the [bug report](https://gitlab.gnome.org/GNOME/pango/-/issues/718) in pango.
The problem seems to be that they moved the Windows backend from GDI to DirectWrite and now `font-family` names that contain font face names, like `Segoe UI Light` or `Roboto Light` are not recognized, so pango falls back on other fonts, which might not support cyrillic as a side effect.

See the last reply I got 
![immagine](https://user-images.githubusercontent.com/43290988/207149611-7f32d9df-56e2-4d65-85af-eeb449b15a62.png)

@Nilvus What do you think? I am not an expert in css and I don't know what to say, is what they propose a viable solution?
Can you help me to reply?

**Comment by Nilvus** [comment]
Thanks @Mark-64 for investigating that. I didn't know that CSS font spec. But make sense when we talk about font-family, not font-name. Will check and test that on my side. Seems not only a viable solution but the good way to code CSS for that as it's in font spec. So seems that this Pango "regression" is not a regression but show not well coded parts.

I just need to make change on my side and be sure there's no downside. Not all CSS specs are supported by Gtk (or not completely) but that one should be ok. Just wait a little, I will test that quickly and answer on Pango issue.

I fear that this CSS update will not be in 4.2 as code freeze is tomorrow or in 2 days and that will not allow time to be sure that CSS change will not introduce another regression.
So, for 4.2, I think it's safe to build Windows package as suggested here with the Pango release that works well with actual CSS code and update CSS code for 4.2.1 and then build Windows package with latest Pango release.

**Comment by wpferguson** [comment]
Pascal just emailed freeze is in 20 hours.

**Comment by Nilvus** [comment]
PR to fix that is posted (#13143): reviews are needed. Thanks guys for team work!

**Comment by TurboGit** [comment]
@Nilvus : Any news on this?

**Comment by Nilvus** [comment]
@TurboGit: unfortunately, no. There's some WIP on Pango side but seems that this issue is also linked to Cairo side. See last comments on https://gitlab.gnome.org/GNOME/pango/-/issues/718 from Chun-Wei Fan And Luca Bacci. MR on Pango Gitlab is https://gitlab.gnome.org/GNOME/pango/-/issues/720

I fear that this will take some time to see a fix on Pango/Cairo side. So, at least for 4.2.1, I think we will need to keep same Pango release on Windows package as for 4.2 (@wpferguson).

**Comment by Nilvus** [comment]
Nothing moves on Pango and Cairo issues. So remove milestone.

**Comment by victoryforce** [comment]
> Note that indeed [removing the offending "Light"](https://gitlab.gnome.org/GNOME/pango/-/issues/718#note_1615502) from the CSS starts addressing the issue...

Yes, this is a fix I've been testing extensively since yesterday (and started thinking about whether it's a good solution even earlier) and will push PR with this fix soon.

I haven't noticed any side effects yet. See, there are themes that do not use "Segoe UI Light", so using this font in the theme is definitely not necessary for correct rendering.
